// powercord/components/ContextMenu/Checkbox.jsx
const { AsyncComponent } = require('powercord/components');
const { React, getAllModules } = require('powercord/webpack');

const Base = require('./Base');
const Checkbox = AsyncComponent.from(getAllModules(m => m.displayName === 'Checkbox')[1]);

module.exports = class CheckboxItem extends Base {
  constructor (props) {
    super(props);

    this.state = {
      active: props.defaultState
    };
  }

  handleOnClick () {
    const active = !this.state.active;
    if (this.props.onToggle) {
      this.props.onToggle(active);
    }

    this.setState({ active });
  }

  render () {
    const { props } = this;
    this.component = React.createElement('div', {
      className: this.filter([ this.classes.item, this.classes.itemToggle, this.classes[!props.disabled ? 'clickable' : 'disabled'] ]),
      style: props.style,
      onClick: this.handleOnClick.bind(this)
    }, React.createElement('span', {
      className: this.classes.label,
      style: props.style
    }, props.name), this.renderCheckbox());

    return super.render();
  }

  renderCheckbox () {
    return React.createElement(Checkbox, {
      checked: this.state.active,
      containerClassName: this.classes.checkbox,
      onChange: () => void 0,
      style: this.props.checkboxStyle
    });
  }
};
