// powercord/constants.js
const { join } = require('path');

/**
 * @module powercord/constants
 */

module.exports = {
  // Powercord
  WEBSITE: 'https://powercord.dev',
  I18N_WEBSITE: 'https://i18n.powercord.dev/projects/powercord',
  REPO_URL: 'vizality/vizality',

  // Runtime
  SETTINGS_FOLDER: join(__dirname, '..', '..', '..', 'settings'),
  CACHE_FOLDER: join(__dirname, '..', '..', '..', '.cache'),
  LOGS_FOLDER: join(__dirname, '..', '..', '..', '.logs'),

  // Discord Server
  DISCORD_INVITE: '42B8AC9',
  GUILD_ID: '689933814864150552',
  MAGIC_CHANNELS: {
    CSS_SNIPPETS: '542864483543023646',
    STORE_PLUGINS: '700461738004578334',
    STORE_THEMES: '700461710972157954',
    STAFF_LOUNGE: '690452551233175602'
  },

  // Flux related
  FluxActions: {
    Settings: {
      LOAD_SETTINGS: 'POWERCORD_SETTINGS_LOAD',
      UPDATE_SETTINGS: 'POWERCORD_SETTINGS_UPDATE',
      TOGGLE_SETTING: 'POWERCORD_SETTING_TOGGLE',
      UPDATE_SETTING: 'POWERCORD_SETTING_UPDATE',
      DELETE_SETTING: 'POWERCORD_SETTING_DELETE'
    },
    Updater: { // @todo: use this
      ADD_UPDATES: 'POWERCORD_UPDATER_ADD',
      UPDATER_CHECKING_BEGIN: 'POWERCORD_UPDATER_CHECKING_BEGIN',
      UPDATER_CHECKING_PROGRESS: 'POWERCORD_UPDATER_CHECKING_PROGRESS',
      UPDATER_CHECKING_SUCCESS: 'POWERCORD_UPDATER_CHECKING_SUCCESS',
      UPDATER_CHECKING_FAILED: 'POWERCORD_UPDATER_CHECKING_FAILED',
      UPDATE_INSTALL_BEGIN: 'POWERCORD_UPDATER_INSTALL_BEGIN',
      UPDATE_INSTALL_PROGRESS: 'POWERCORD_UPDATER_INSTALL_PROGRESS',
      UPDATE_INSTALL_SUCCESS: 'POWERCORD_UPDATER_INSTALL_SUCCESS',
      UPDATE_INSTALL_FAILED: 'POWERCORD_UPDATER_INSTALL_FAILED'
    }
  }
};
